### ISSUE TITLE
many0!(...) is not equal to separated_list!(tag!(""), ...)

### ISSUE THREAD

**Issue by tstorch** [Geal/nom#177]
This took me some time to understand, but I now think why that is. To get a hint try to change in current master 704f508 in marcros.rs line 1865

``` rust
if ($i).input_len() == 0 {
  $crate::IResult::Done($i, ::std::vec::Vec::new())
}
```

with

``` rust
if ($i).input_len() == 0 {
  return $crate::IResult::Done($i, ::std::vec::Vec::new());
}
```

I thought this would change nothing, but it did. The fact that this is a macro means, that it is placed verbatim, where it's called. This is the reason, why this expression does not always lead the "calling function" to return this. In separated_list return is used instead of expression return.

@Geal Are you aware of this? Should it be this way? should I change separated_list to mimic this?

**Comment by Geal** [comment]
Hmm, I did not use `return` in the previous `separated_list!` implementation, this is something that you introduced in 5715db59da4 and that I did not see in the code review.

I'm sorry, but I'll have to revert your PR until I get some time to fix things (or you find a way to make it work again).

I would also appreciate that you ask me and wait a bit before refactoring everything in nom. While you may think that assembling the implementations of `many0!` and `separated_list!`is good because it reduces the quantity of code, it may break some behaviour that  we don't see yet, and does not necessarily make the code easier to understand.
Please do not take this the wrong way, I really appreciated your help and your enthusiasm so far. It's just that I do not have much time to work on nom and review code this month, for personal and professional reasons.

By the way, could you tell me more about your use cases for nom parsers? It may help me understand the context and what you need here.

**Comment by tstorch** [comment]
I now that I introduced them :stuck_out_tongue_winking_eye:. I wasn't aware of this...

I didn't restructure anything yet. My coment about refactoring was just to ask for opinions.

Please do not revert the commit as this is fixed in PR #178

**Comment by Geal** [comment]
Ah, sorry, too late, I already reverted it. Could you provide a separate PR for the `separated_list!` fix without early returns?

**Comment by tstorch** [comment]
Later this week...

My background is, that I wanted to write a small SQL parser with nom and I stumble across the flaw in alpha, digit etc. Out of pure interest I browsed your code and found it at parts hard to read, so I rewrote some parts and submitted the PRs (I dig readable code).

I do not take this the wrong way! I'm very glad that someone (you) is providing a this awsome lib. 

Regarding my PRs: From time to time I get overenthusiastic and things get messy :sweat_smile:
